lib/faraday/request/multipart_with_file.rb

```diff
@@ -5,9 +5,15 @@ module Faraday
   # @private
   class Request::MultipartWithFile < Faraday::Middleware
     def call(env)
+
+      # Faraday seems to expect a few IO methods to be available, but that's all:
+      # https://github.com/lostisland/faraday/blob/master/lib/faraday/file_part.rb
+      # :length seems to be an optional one
+      file_like_objects_respond_to = ["rewind", "read", "close"].map(&:to_sym)
+
       if env[:body].is_a?(Hash)
         env[:body].each do |key, value|
-          if value.is_a?(File)
+          if file_like_objects_respond_to.all? { |m| value.respond_to? (m) }
```

Comment:
Looking at the docs/code for `Faraday::FilePart` and `UploadIO` from `multipart-post`, I think the expectation is just a standard `IO` interface (or a string filename). I'd prefer to match that in the condition, instead of duck-typing.

Can you please try the following to see if it works for your use-case?

```suggestion
          if value.is_a?(IO)
```

Reply:
Replacing `File` with `IO` sounds great to me.  That would have been my first choice for the entire PR, but it fails in my particular use case.

`Zip::InputStream` isn't an `IO`, it's an ["IO like object" ](https://www.rubydoc.info/github/rubyzip/rubyzip/Zip/Entry#get_input_stream-instance_method).  Put simply:
```
irb(main):001:0> require 'zip'
=> true
irb(main):002:0> Zip::InputStream.new("foo").is_a? IO
=> false
irb(main):003:0> ["rewind", "read", "close"].map(&:to_sym).all? { |m| Zip::InputStream.new("foo").respond_to? m }
=> true
```

I also considered submitting a change to the Faraday library, but I couldn't find any place where they were making an explicit check for an `IO` object -- they were just assuming those methods existed.  Similarly, [the source of `UploadIO` (click to expand)](https://www.rubydoc.info/gems/multipart-post/2.0.0/UploadIO:initialize) does a `respond_to?` check, not an `is_a?` check.  Hence the duck typing here -- they'll allow it if you allow it.

Reply:
You're right - thank you for clarifying, and for putting it in writing.

Late note: to avoid cognitive complexity, and extra allocations, I'd just be explicit here with:

``` ruby
if value.respond_to?(:read) && value.respond_to?(:rewind) && value.respond_to?(:close)
```

And that condition covers `value.is_a?(IO)`, so no need to have both.

Reply:
Works for me.  I pulled that out into an array to make it easer to attach the comments to that, but easy enough to put that above the `if`.  I'll make that change now.